c/force_reconfiguration: may the force (reconfiguration) be with you

bharathv opened the pull request:
Allows force reconfiguration of partition in any state (including stuck force reconfiguration). This behavior change helps unblock stuck partitions due to the following reasons (for example, stuck in move / cancelation without leader, stuck in force reconfiguration etc) which are difficult to unblock today without a lot of gymnastics.

Fixes: https://redpandadata.atlassian.net/browse/INC-831

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
### Features

* Force reconfiguration now supports updating any in-progress move, regardless of its state. This makes it much easier to use as an escape hatch to recover partitions in various stuck scenarios that would otherwise be difficult to recover.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
c/topics: add update_final_counts() util

no logical changes

bharathv committed:
c/topic_table: add update_in_progress(ntp&) utility

bharathv committed:
c/force_reconfiguration: support force reconfiguring in progress moves

bharathv committed:
c/topic_table: remove unusued variable

bharathv committed:
c/frontend: relax force reconfiguration requirements

bharathv commented:
/dt

bharathv commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--48cdb048-0797-4e17-8f8a-16de26aa3ca9-->
<!--TEST RESULTS-->
<!--445e1c68-5b35-4abe-aff6-08a950e293bd-->
<!--TEST RESULTS-->
<!--450814f5-0c42-435a-ae65-b6a2f5ef4ea1-->
<!--TEST RESULTS-->
<!--2dfbd554-ad3c-4be8-b407-faafc545597f-->
<!--TEST RESULTS-->
<!--83ef3d49-0d75-4d47-8a6d-818cc8254351-->
<!--TEST RESULTS-->
<!--f6b8c97b-1ece-4cf6-a015-f0db738d6c00-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67949">build#67949</a></summary>


| test_class                | test_method                           | test_arguments                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:--------------------------|:--------------------------------------|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionMoveInterruption | test_cancelling_partition_move_x_core | {"compacted": false, "recovery": "restart_recovery", "replication_factor": 3, "unclean_abort": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67949#0197aa8c-0f7c-4ab4-82a9-23bc9eedaae6 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
| SimpleEndToEndTest        | test_relaxed_acks                     | {"write_caching": false}                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67949#0197aa8c-828a-4b68-ba35-0148441ad5f3 | FLAKY         | 20/21    | upstream reliability is '99.02439024390245'. current run reliability is '95.23809523809523'. drift is 3.7863 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68033">build#68033</a></summary>


| test_class                        | test_method                              | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------------|:-----------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EnterpriseFeaturesTest            | test_enable_features                     | {"disable_trial": false, "feature": 5, "install_license": true}                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68033#0197aec5-1ea7-4cd5-a9cb-ce014fcdb376 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionForceReconfigurationTest | test_reconfiguring_force_reconfiguration |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68033#0197aec1-e070-4ddd-9241-15bcf307a7fb | FLAKY         | 13/21    | The test was found to be new, and no failures are allowed                                                                                                              |
| RandomNodeOperationsTest          | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68033#0197aec1-e06f-4529-909c-88e9c4487a36 | FLAKY         | 19/21    | upstream reliability is '98.6413043478261'. current run reliability is '90.47619047619048'. drift is 8.16511 and the allowed drift is set to 50. The test should PASS  |
| RecreateTopicMetadataTest         | test_recreated_topic_metadata_are_valid  | {"replication_factor": 3}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68033#0197aec1-e071-47fc-bba3-e089508f118d | FLAKY         | 20/21    | upstream reliability is '98.22616407982262'. current run reliability is '95.23809523809523'. drift is 2.98807 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest       | drop_lifecycle_marker_test               | {"cloud_storage_type": 1}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68033#0197aec5-1ea7-4cd5-a9cb-ce014fcdb376 | FLAKY         | 20/21    | upstream reliability is '96.69724770642202'. current run reliability is '95.23809523809523'. drift is 1.45915 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68140">build#68140</a></summary>


| test_class                     | test_method              | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------------|:-------------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeCustomPartitioningTest | test_spec_evolution_rpcn | {"catalog_type": "nessie", "cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68140#0197b4e4-b439-4d07-abcb-c311278bd705 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68150">build#68150</a></summary>


| test_class                        | test_method                           | test_arguments                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------------|:--------------------------------------|:---------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionForceReconfigurationTest | test_basic_reconfiguration            | {"acks": -1, "controller_snapshots": false, "restart": false}                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa34-4e75-8447-7b3c62fa73cd | FLAKY         | 69/70    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionForceReconfigurationTest | test_basic_reconfiguration            | {"acks": 1, "controller_snapshots": false, "restart": false}                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa39-4da0-b499-7dc6136d77ff | FLAKY         | 69/70    | upstream reliability is '99.69788519637463'. current run reliability is '95.23809523809523'. drift is 4.45979 and the allowed drift is set to 50. The test should PASS |
| PartitionForceReconfigurationTest | test_basic_reconfiguration            | {"acks": 1, "controller_snapshots": false, "restart": true}                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa3b-460c-a677-54f24c45f623 | FLAKY         | 69/70    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionForceReconfigurationTest | test_reconfiguring_in_progress_move   | {"cancellation": false}                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa36-40cb-9515-456fe494aecf | FLAKY         | 69/70    | The test was found to be new, and no failures are allowed                                                                                                              |
| PartitionMoveInterruption         | test_cancelling_all_moves_from_node   |                                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa3a-4573-89d4-8b1276362e4d | FLAKY         | 69/70    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionMoveInterruption         | test_cancelling_partition_move        | {"compacted": false, "force_back": false, "recovery": "restart_recovery", "replication_factor": 1, "unclean_abort": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa34-43d6-972f-652c21f83d12 | FLAKY         | 69/70    | The test was found to be new, and no failures are allowed                                                                                                              |
| PartitionMoveInterruption         | test_cancelling_partition_move        | {"compacted": false, "force_back": true, "recovery": "restart_recovery", "replication_factor": 1, "unclean_abort": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa38-4025-ad5e-c104100a510c | FLAKY         | 69/70    | The test was found to be new, and no failures are allowed                                                                                                              |
| PartitionMoveInterruption         | test_cancelling_partition_move        | {"compacted": false, "force_back": true, "recovery": "no_recovery", "replication_factor": 1, "unclean_abort": true}        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa33-4f58-ad43-1893082bcecf | FLAKY         | 69/70    | The test was found to be new, and no failures are allowed                                                                                                              |
| PartitionMoveInterruption         | test_cancelling_partition_move        | {"compacted": false, "force_back": false, "recovery": "restart_recovery", "replication_factor": 3, "unclean_abort": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa3a-4573-89d4-8b1276362e4d | FLAKY         | 69/70    | upstream reliability is 'nan'. current run reliability is '95.23809523809523'. drift is nan and the allowed drift is set to 50. The test should FAIL                   |
| PartitionMoveInterruption         | test_cancelling_partition_move        | {"compacted": false, "force_back": false, "recovery": "restart_recovery", "replication_factor": 3, "unclean_abort": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa3a-463f-b369-937f536f0f7c | FLAKY         | 69/70    | The test was found to be new, and no failures are allowed                                                                                                              |
| PartitionMoveInterruption         | test_cancelling_partition_move_x_core | {"compacted": false, "recovery": "restart_recovery", "replication_factor": 1, "unclean_abort": false}                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa37-4bc2-b98a-76513a9353f7 | FLAKY         | 69/70    | upstream reliability is '99.69788519637463'. current run reliability is '95.23809523809523'. drift is 4.45979 and the allowed drift is set to 50. The test should PASS |
| PartitionMovementTest             | test_movement_tracking_api            |                                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68150#0197b7d9-aa35-45e7-b3da-887600f03039 | FLAKY         | 69/70    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68234">build#68234</a></summary>


| test_class                | test_method                                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:----------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_when_segment_size_smaller_than_chunk_size |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68234#0197c426-535d-4b36-bc33-38e014700488 | FLAKY         | 20/21    | upstream reliability is '98.54721549636804'. current run reliability is '95.23809523809523'. drift is 3.30912 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68328">build#68328</a></summary>


| test_class        | test_method                               | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------|:------------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest      | test_single_partition_leadership_transfer | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68328#0197cba8-7855-4fe6-b128-c525939b8edd | FLAKY         | 20/21    | upstream reliability is '99.72602739726028'. current run reliability is '95.23809523809523'. drift is 4.48793 and the allowed drift is set to 50. The test should PASS  |
| ClusterConfigTest | test_valid_settings                       |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68328#0197cbc4-b63a-4959-a1de-9612823c1442 | FLAKY         | 18/21    | upstream reliability is '83.86075949367088'. current run reliability is '85.71428571428571'. drift is -1.85353 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

bharathv committed:
dt/tests: split reconfiguration utility

Adds normal and force reconfiguration methods using the same base
implementation.

bharathv committed:
dt/tests: tests for force reconfiguration with in progress moves

bharathv committed:
dt/tests: wireup partition movement tests with force reconfiguration

bharathv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68033-->
<!--49a597c2-c8fb-4884-9de0-8e5a626579b0-->
<!--title-->
## Retry command for [Build#68033](https://buildkite.com/redpanda/redpanda/builds/68033)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_force_reconfiguration_test.py::PartitionForceReconfigurationTest.test_reconfiguring_force_reconfiguration
```
<!--/body-->

bharathv commented:
/dt

bharathv commented:
/ci-repeat 3
skip-units
skip-redpanda-build
dt-repeat=100
tests/rptest/tests/datalake/delayed_translation_test.py
tests/rptest/tests/partition_move_interruption_test.py
tests/rptest/tests/partition_force_reconfiguration_test.py
tests/rptest/tests/partition_movement_test.py

bharathv commented:
/ci-repeat 3
skip-units
skip-redpanda-build
dt-repeat=50
tests/rptest/tests/partition_move_interruption_test.py
tests/rptest/tests/partition_force_reconfiguration_test.py
tests/rptest/tests/partition_movement_test.py

bharathv committed:
c/tests: add tests for allocator changes with force reconfiguraiton

mmaslankaprv commented on src/v/cluster/topic_updates_dispatcher.cc:
```diff
@@ -433,7 +430,77 @@ ss::future<std::error_code> topic_updates_dispatcher::apply(
+        //  to_remove = [4] A: [1:1, 2:1, 3:1, 4:0]
+        //  -- step 2 -- add allocations for [4, 6, 8] -- to_add = [4:1, 6:1,
+        //  8:1]  A: [1:1, 2:1, 3:1, 4:1, 6:1, 8:1]
+        //  -- step 3 -- update final counts : F [1:0, 2:0, 4:1, 6:1, 8:1]
```
is this missing `3:0` ?

bharathv replied:
oops yes, added it.

mmaslankaprv commented:
/ci-repeat 3
skip-units
skip-redpanda-build
dt-repeat=50
tests/rptest/tests/partition_move_interruption_test.py
tests/rptest/tests/partition_force_reconfiguration_test.py
tests/rptest/tests/partition_movement_test.py

vbotbuildovich commented:
/backport v25.1.x

dotnwat commented on src/v/cluster/topic_updates_dispatcher.cc:
```diff
@@ -433,7 +430,77 @@ ss::future<std::error_code> topic_updates_dispatcher::apply(
+        // undo any allocations from in progress move.
+
+        // For example, consider the following move
+        // --- (A (prev) -> B (target)) -> C (force)
+
+        // During the move A -> B, allocations amounting to (B - A) are added
+        // to account for the in progress move. Here we undo those allocations.
+        //
+        // note: A cancellation only updates final moves, so no allocator
+        // changes are made.
+
+        // Detailed example:
+
+        // Say a partition is being force reconfigured from [1, 2, 3] -> [2, 3,
+        // 4] and then force reconfigured to [4, 6, 8] while the reconfiguration
+        // is stuck.
+
+        // To make it more interesting, say there was a cancellation of the
+        // reconfiguration which is stuck too because [2, 3] are not available.
+
+        // Initial state:
+        // allocations (A) and final counts (F) follow {broker: allocations}
+        // dict notation.
+
+        // [1, 2, 3], allocations: [1:1, 2:1, 3:1], final counts: [1:1, 2:1,
+        // 3:1]
+        //
+        // reconfiguration [1, 2, 3] -> [2, 3, 4], A: [1:1, 2:1, 3:1, 4:1], F:
+        // [1:0, 2:1, 3:1, 4:1]
+        //
+        // reconfiguration is stuck and canceled
+        //
+        // reconfiguration [1, 2, 3] <- (cancel) [2, 3, 4] A: [1:1, 2:1, 3:1,
+        // 4:1], F: [1:1, 2:1, 3:1, 4:0]
+
+        // force reconfiguration to [4, 6, 8] while the canceling
+        // reconfiguration is stuck
+
+        // force reconfiguration [1, 2, 3] -> [4, 6, 8]
+        //  -- step 1 -- undo allocations from [1, 2, 3] -> [2, 3, 4] --
```
🙇

daisukebe replied:
Thank you for describing how it works clearly!
